File: Sources/SkillManagerCore/InventoryService.swift
import Foundation

public final class InventoryService: @unchecked Sendable {
    private let scanner: SkillScanner
    private let usageAnalyzer: UsageAnalyzer
    private let archiveStore: ArchiveStore
    private let historyStore: OperationHistoryStore
    private let reportStore: CleanupReportStore
    private let decisionStore: SkillDecisionStore

    public init(
        scanner: SkillScanner = SkillScanner(),
        usageAnalyzer: UsageAnalyzer = UsageAnalyzer(),
        archiveStore: ArchiveStore = ArchiveStore(),
        historyStore: OperationHistoryStore = OperationHistoryStore(),
        reportStore: CleanupReportStore = CleanupReportStore(),
        decisionStore: SkillDecisionStore = SkillDecisionStore()
    ) {
        self.scanner = scanner
        self.usageAnalyzer = usageAnalyzer
        self.archiveStore = archiveStore
        self.historyStore = historyStore
        self.reportStore = reportStore
        self.decisionStore = decisionStore
    }

    public func loadInventory(now: Date = Date()) -> SkillInventory {
        let roots = scanner.defaultRoots()
        let roughSkills = scanner.scan(roots: roots, usage: [:], now: now)
        let usage = usageAnalyzer.analyzeSkillUsage(skills: roughSkills)
        let skills = scanner.scan(roots: roots, usage: usage, now: now)
        return SkillInventory(
            active: skills,
            archived: archiveStore.archivedSkills(),
            scannedAt: now
        )
    }

    public func defaultRoots() -> [SkillRoot] {
        scanner.defaultRoots()
    }

    public func auditReport(for inventory: SkillInventory, generatedAt: Date = Date()) -> InventoryAuditReport {
        inventory.auditReport(roots: scanner.defaultRoots(), generatedAt: generatedAt)
    }

    public func sessionRootAudits() -> [UsageSessionRootAudit] {
        usageAnalyzer.sessionRootAudits()
    }

    public func operationHistory() -> [SkillOperationEntry] {
        historyStore.entries()
    }

    public func skillDecisions() -> [String: SkillDecisionRecord] {
        decisionStore.decisions()
    }

    public func setDecision(_ decision: SkillUserDecision?, for skill: SkillRecord, now: Date = Date()) throws {
        try decisionStore.setDecision(decision, for: skill.id, now: now)
    }

    public func decision(for skill: SkillRecord) -> SkillUserDecision? {
        decisionStore.decision(for: skill.id)
    }

    public var cleanupReportsDirectoryURL: URL {
        reportStore.reportsRootURL
    }

    public func cleanupCandidates(in inventory: SkillInventory) -> [SkillRecord] {
        let decisions = decisionStore.decisions()
        return inventory.archiveCandidates.filter { skill in
            guard let decision = decisions[skill.id]?.decision else { return true }
            return decision != .protected && decision != .review
        }
    }

    public func protectedSkills(in inventory: SkillInventory) -> [SkillRecord] {
        let decisions = decisionStore.decisions()
        return inventory.active.filter { decisions[$0.id]?.decision == .protected }
    }

    public func reviewSkills(in inventory: SkillInventory) -> [SkillRecord] {
        let decisions = decisionStore.decisions()
        return inventory.active.filter { skill in
            if decisions[skill.id]?.decision == .protected { return false }
            return decisions[skill.id]?.decision == .review || skill.recommendation == .review
        }
    }

    public func exportCleanupReport(inventory: SkillInventory, skills: [SkillRecord], now: Date = Date()) throws -> CleanupReportExport {
        try reportStore.export(inventory: inventory, skills: skills, decisions: decisionStore.decisions(), now: now)
    }

    @discardableResult
    public func archive(_ skill: SkillRecord) throws -> ArchivedSkill {
        do {
            let archived = try archiveStore.archive(skill)
            try? historyStore.append(SkillOperationEntry(
                action: .archive,
                skillName: skill.name,
                title: skill.title,
                originalPath: skill.path,
                archivePath: archived.archivePath,
                succeeded: true,
                message: "Archived"
            ))
            return archived
        } catch {
            try? historyStore.append(SkillOperationEntry(
                action: .archive,
                skillName: skill.name,
                title: skill.title,
                originalPath: skill.path,
                archivePath: nil,
                succeeded: false,
                message: error.localizedDescription
            ))
            throw error
        }
    }

    public func restore(_ archived: ArchivedSkill) throws {
        do {
            try archiveStore.restore(archived)
            try? historyStore.append(SkillOperationEntry(
                action: .restore,
                skillName: archived.name,
                title: archived.title,
                originalPath: archived.originalPath,
                archivePath: archived.archivePath,
                succeeded: true,
                message: "Restored"
            ))
        } catch {
            try? historyStore.append(SkillOperationEntry(
                action: .restore,
                skillName: archived.name,
                title: archived.title,
                originalPath: archived.originalPath,
                archivePath: archived.archivePath,
                succeeded: false,
                message: error.localizedDescription
            ))
            throw error
        }
    }
}

File: packages/core/src/inventory-service.ts
import path from "node:path";

import { ArchiveStore } from "./archive-store.js";
import { SkillDecisionStore } from "./decision-store.js";
import { CleanupReportStore } from "./report-store.js";
import { SkillScanner } from "./scanner.js";
import { UsageAnalyzer } from "./usage-analyzer.js";
import type { ArchivedSkill, CleanupReportExport, InventoryAuditReport, InventoryServiceOptions, SkillDecision, SkillDecisionRecord, SkillInventory, SkillRecord } from "./types.js";

export class InventoryService {
  readonly homeDir: string;
  readonly userDataDir: string;
  readonly scanner: SkillScanner;
  readonly usageAnalyzer: UsageAnalyzer;
  readonly archiveStore: ArchiveStore;
  readonly decisionStore: SkillDecisionStore;
  readonly reportStore: CleanupReportStore;

  constructor(options: Required<Pick<InventoryServiceOptions, "homeDir" | "userDataDir">> & InventoryServiceOptions) {
    this.homeDir = options.homeDir;
    this.userDataDir = options.userDataDir;
    this.scanner = new SkillScanner({ homeDir: this.homeDir });
    this.usageAnalyzer = new UsageAnalyzer({
      homeDir: this.homeDir,
      maxLogBytes: options.maxLogBytes,
      maxLogFiles: options.maxLogFiles
    });
    this.archiveStore = new ArchiveStore(this.userDataDir);
    this.decisionStore = new SkillDecisionStore(this.userDataDir);
    this.reportStore = new CleanupReportStore(this.userDataDir);
  }

  async loadInventory(now = new Date()): Promise<SkillInventory> {
    const roots = this.scanner.defaultRoots();
    const roughSkills = await this.scanner.scan(roots, new Map(), now);
    const usage = await this.usageAnalyzer.analyzeSkillUsage(roughSkills);
    const usageWarnings = this.usageAnalyzer.warningsBySkillId();
    const active = (await this.scanner.scan(roots, usage, now)).map((skill) => ({
      ...skill,
      scanWarnings: [...skill.scanWarnings, ...(usageWarnings.get(skill.id) ?? [])]
    }));
    const decisions = await this.decisionStore.all();
    const activeWithDecisions = active.map((skill) => applyDecision(skill, decisions.get(skill.id)));
    const archived = await this.archiveStore.archivedSkills();
    const audit = await this.auditReport(activeWithDecisions, archived, now);
    const sessionRootAudits = await this.usageAnalyzer.sessionRootAudits();
    return {
      active: activeWithDecisions,
      archived,
      scannedAt: now.toISOString(),
      audit,
      sessionRootAudits,
      usageEvidenceAudit: this.usageAnalyzer.analysisAudit()
    };
  }

  async setDecision(skillId: string, decision: SkillDecision | null): Promise<void> {
    await this.decisionStore.set(skillId, decision);
  }

  async archiveSkill(skill: SkillRecord): Promise<ArchivedSkill> {
    return this.archiveStore.archive(skill);
  }

  async archiveSkillById(skillId: string): Promise<ArchivedSkill> {
    const inventory = await this.loadInventory();
    const skill = inventory.active.find((item) => item.id === skillId);
    if (!skill) throw new Error(`Unknown active skill: ${skillId}`);
    return this.archiveStore.archive(skill);
  }

  async restoreSkill(archived: ArchivedSkill): Promise<ArchivedSkill> {
    return this.archiveStore.restore(archived);
  }

  async restoreArchivedById(archivedId: string): Promise<ArchivedSkill> {
    const inventory = await this.loadInventory();
    const archived = inventory.archived.find((item) => item.id === archivedId);
    if (!archived) throw new Error(`Unknown archived skill: ${archivedId}`);
    return this.archiveStore.restore(archived);
  }

  async exportCleanupReport(inventory: SkillInventory, skills: SkillRecord[]): Promise<CleanupReportExport> {
    return this.reportStore.export(inventory, skills, await this.decisionStore.all());
  }

  async revealPathTarget(targetPath: string): Promise<string> {
    return path.resolve(targetPath);
  }

  private async auditReport(active: SkillRecord[], archived: ArchivedSkill[], now: Date): Promise<InventoryAuditReport> {
    const roots = this.scanner.defaultRoots();
    const rootAudits = await this.scanner.auditRoots(roots, active);
    const archiveCandidates = active.filter((skill) => skill.recommendation === "archive" && !skill.status.protected && !skill.status.reviewLater);
    return {
      generatedAt: now.toISOString(),
      installedCount: active.length,
      archivedCount: archived.length,
      unusedCount: active.filter((skill) => skill.usageCount === 0).length,
      suggestedArchiveCount: archiveCandidates.length,
      contextTokens: active.reduce((sum, skill) => sum + skill.contextTokens, 0),
      reclaimableContextTokens: archiveCandidates.reduce((sum, skill) => sum + skill.contextTokens, 0),
      reclaimableBytes: archiveCandidates.reduce((sum, skill) => sum + skill.sizeBytes, 0),
      roots: rootAudits
    };
  }
}

function applyDecision(skill: SkillRecord, decision: SkillDecisionRecord | undefined): SkillRecord {
  if (!decision) return skill;
  const status = { ...skill.status };
  if (decision.decision === "protected") {
    status.protected = true;
  }
  if (decision.decision === "review") {
    status.reviewLater = true;
  }
  return {
    ...skill,
    status,
    recommendation: decision.decision === "protected" || decision.decision === "review" ? "keep" : skill.recommendation
  };
}
